client-local/src/main/java/com/cloudera/livy/client/local/ContextLauncher.java

```diff
@@ -0,0 +1,446 @@
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import io.netty.channel.ChannelHandlerContext;
+import org.apache.spark.launcher.SparkLauncher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.cloudera.livy.client.local.driver.RemoteDriver;
+import com.cloudera.livy.client.local.rpc.Rpc;
+import com.cloudera.livy.client.local.rpc.RpcDispatcher;
+import com.cloudera.livy.client.local.rpc.RpcServer;
+import static com.cloudera.livy.client.local.LocalConf.Entry.*;
+
+/**
+ * Encapsulates code needed to launch a new Spark context and collect information about how
+ * to establish a client connection to it.
+ */
+class ContextLauncher implements ContextInfo {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ContextLauncher.class);
+  private static final AtomicInteger CHILD_IDS = new AtomicInteger();
+
+  private static final String SPARK_JARS_KEY = "spark.jars";
+  private static final String SPARK_HOME_ENV = "SPARK_HOME";
+  private static final String SPARK_HOME_KEY = "spark.home";
+
+  private final String clientId;
+  private final String secret;
+  private final ChildProcess child;
+  private final LocalConf conf;
+  private final LocalClientFactory factory;
+
+  private BaseProtocol.RemoteDriverAddress driverAddress;
+
+  ContextLauncher(LocalClientFactory factory, LocalConf conf) throws IOException {
+    this.clientId = UUID.randomUUID().toString();
+    this.secret = factory.getServer().createSecret();
+    this.conf = conf;
+    this.factory = factory;
+
+    RegistrationHandler handler = new RegistrationHandler();
+    try {
+      factory.getServer().registerClient(clientId, secret, handler);
+      this.child = startDriver(factory.getServer(), conf, clientId, secret);
+
+      // Wait for the handler to receive the driver information. Wait a little at a time so
+      // that we can check whether the child process is still alive, and throw an error if the
+      // process goes away before we get the information back.
+      long timeout = conf.getTimeAsMs(RPC_CLIENT_HANDSHAKE_TIMEOUT);
+      long deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout);
+      long now = System.nanoTime();
+      long step = Math.min(500, TimeUnit.NANOSECONDS.toMillis(deadline - now));
+      synchronized (handler) {
+        while (handler.driverAddress == null && now < deadline) {
+          handler.wait(Math.min(step, deadline - now));
+          Preconditions.checkState(handler.driverAddress != null || !child.isFailed(),
+            "Child has exited before address information was received.");
+          now = System.nanoTime();
+        }
+      }
+      Preconditions.checkState(handler.driverAddress != null,
+        "Timed out waiting for driver connection information.");
+      driverAddress = handler.driverAddress;
+    } catch (Exception e) {
+      factory.getServer().unregisterClient(clientId);
+      throw Throwables.propagate(e);
```

Comment:
This can just be `Throwables.propogate(e)`. The `throw` is required only if the calling method expects a return value.